corehq/apps/domain/deletion.py

```diff
@@ -0,0 +1,103 @@
+from django.apps import apps
+from django.core.exceptions import ImproperlyConfigured
+from django.db import connection
+
+
+class BaseDeletion(object):
+    def __init__(self, app_label):
+        self.app_label = app_label
+
+    def is_app_installed(self):
+        try:
+            return apps.get_app(self.app_label)
+        except ImproperlyConfigured:
+            return False
+
+
+class CustomDeletion(BaseDeletion):
+    def __init__(self, app_label, deletion_fn):
+        super(CustomDeletion, self).__init__(app_label)
+        self.deletion_fn = deletion_fn
+
+    def execute(self, domain_name):
+        if self.is_app_installed():
+            self.deletion_fn(domain_name)
+
+
+class RawDeletion(BaseDeletion):
+    def __init__(self, app_label, raw_query):
+        super(RawDeletion, self).__init__(app_label)
+        self.raw_query = raw_query
+
+    def execute(self, cursor, domain_name):
+        if self.is_app_installed():
+            cursor.execute(self.raw_query, [domain_name])
+
+
+class ModelDeletion(BaseDeletion):
+    def __init__(self, app_label, model_name, domain_filter_kwarg):
+        super(ModelDeletion, self).__init__(app_label)
+        self.domain_filter_kwarg = domain_filter_kwarg
+        self.model_name = model_name
+
+    def get_model_class(self):
+        return apps.get_model(self.app_label, self.model_name)
```

Comment:
This will raise an exception if the model can't be found. Intentionally not catching that since that probably means there's a typo in the model name. This only gets called if the app is installed.